lib/spack/spack/build_systems/intel.py

```diff
@@ -103,6 +136,742 @@ def _filtered_components(self):
+    # --------------------------------------------------------------------
+    #
+    # This is a product bundle whose install dir mostly holds merely symlinks
+    # to components installed in sibling dirs:i
+    #
+    #   $ ls -lF /opt/intel/parallel_studio_xe_2018.1.038/
+    #   advisor_2018		 -> /opt/intel/advisor_2018/
+    #   clck_2018		 -> /opt/intel/clck/2018.1/
+    #   compilers_and_libraries_2018 -> /opt/intel/comp....aries_2018.1.163/
+    #   documentation_2018	 -> /opt/intel/documentation_2018/
+    #   ide_support_2018	 -> /opt/intel/ide_support_2018/
+    #   inspector_2018		 -> /opt/intel/inspector_2018/
+    #   itac_2018		 -> /opt/intel/itac/2018.1.017/
+    #   man		         -> /opt/intel/man/
+    #   samples_2018		 -> /opt/intel/samples_2018/
+    #   vtune_amplifier_2018	 -> /opt/intel/vtune_amplifier_2018/
+    #
+    #   psxevars.csh		 -> ./bin/psxevars.csh*
+    #   psxevars.sh		 -> ./bin/psxevars.sh*
+    #   bin/            - *vars.*sh scripts + sshconnectivity.exp ONLY
+    #
+    #   licensing/
+    #   uninstall*
+    #
+    # The only relevant regular files are *vars.*sh, but those also just churn
+    # through the subordinate vars files of the components.
+    #
+    # --------------------------------------------------------------------
+    #
+    # Note on macOS support, i.e., sys.platform == 'darwin':
+    #
+    # - On macOS, the Spack methods here only include support to integrate an
+    #   externally installed MKL.
+    #
+    # - URLs in child packages will be Linux-specific; macOS download packages
+    #   are located in differently numbered dirs and are named m_*.dmg.
+    #
+    # --------------------------------------------------------------------
+
+    def product_dir(self, product_dir_name, version_glob='_2???.*.*[0-9]',
+                    postfix_dir=''):
+        '''Returns the version-specific directory of an Intel product release,
+        holding the main product and possibly auxiliary files from other
+        products.
+        '''
+        d = self.prefix
+        if sys.platform == 'darwin':
+            # TODO: Verify on Mac.
+            return d
+
+        # Distinguish between product installations that were done outside of
+        # Spack (as "ghost package" integrated via packages.yaml) and
+        # Spack-native ones. The key issue is that their prefixes will
+        # qualitatively differ in directory depth.
+        if ('%s_' % product_dir_name) in d and '.' in d:
+            # If e.g. MKL was installed outside of Spack, it is likely just one
+            # product or product component among possibly many other Intel
+            # products and their releases that were installed in sibling or
+            # cousin directories.  In such cases, the prefix given to Spack
+            # will inevitably be a highly product-specific and preferably fully
+            # version-specific directory.  This is what we want and need.
```

Comment:
For some reason, I was under the impression that people who ran into these problems were simply trying to use the `intel-mkl` external package when what they had installed was actually `intel-parallel-studio`. I guess I was under the wrong impression.

Reply:
When I realized that semantic difference between a "vendor"-level ``prefix`` in Spack and a "component"-level one, I was wondering: are there other package families that would have similar issues? (Perl, Python, conda, R?)